dali/operators/decoder/audio/libsnd_decoder.h

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_LIBSND_DECODER_H
+#define DALI_LIBSND_DECODER_H
+
+#include <cstring>
+#include <cassert>
+#include <vector>
+#include "dali/core/format.h"
+#include "dali/core/error_handling.h"
+#include "dynlink_snd.h"
+#include "audio_decoder.h"
+
+
+namespace dali {
+
+namespace detail {
+
+class VirtualInputManager {
+ public:
+  VirtualInputManager(const void *input, int length /* bytes */) :
+          length_(length), curr_(0), input_(static_cast<const char *>(input)) {}
+
+
+  size_t length() {
+    return length_;
+  }
+
+
+  size_t seek(sf_count_t offset, int whence) {
+    switch (whence) {
+      case SEEK_SET:
+        curr_ = offset;
+        break;
+      case SEEK_CUR:
+        curr_ += offset;
+        break;
+      case SEEK_END:
+        curr_ = length_ + offset;
+        break;
+      default:
+        DALI_FAIL("Incorrect `whence` argument")
+    }
+    return curr_;
+  }
+
+
+  size_t read(void *dst, sf_count_t num) {
+    num = std::min<sf_count_t>(num, length_ - curr_);
+    memcpy(dst, input_ + curr_, num);
+    curr_ += num;
+    return num;
+  }
+
+
+  size_t tell() {
+    return curr_;
+  }
+
+
+ private:
+  size_t length_, curr_;
+  const char *input_;
+};
+
+
+/**
+ * The virtual file context must return the length of the virtual file in bytes.
+ */
+sf_count_t GetFileLen(void *This) {
+  return static_cast<VirtualInputManager *>(This)->length();
+}
+
+
+/**
+ *  The virtual file context must seek to offset using the seek mode provided by whence
+ *  which is one of
+ *   SEEK_CUR
+ *   SEEK_SET
+ *   SEEK_END
+ * The return value must contain the new offset in the file.
+ */
+sf_count_t Seek(sf_count_t offset, int whence, void *This) {
+  return static_cast<VirtualInputManager *>(This)->seek(offset, whence);
+}
+
+
+/**
+ * The virtual file context must copy ("read") "num" bytes into the buffer provided by dst
+ * and return the count of actually copied bytes.
+ */
+sf_count_t Read(void *dst, sf_count_t num, void *This) {
+  return static_cast<VirtualInputManager *>(This)->read(dst, num);
+}
+
+
+/**
+ * Return the current position of the virtual file context.
+ */
+sf_count_t Tell(void *This) {
+  return static_cast<VirtualInputManager *>(This)->tell();
+}
+
+}  // namespace detail
+
+template<typename T>
+struct LibsndWavDecoder : public WavDecoder<T>, public AllocatingDecoder<T> {
```

Comment:
Do we want this configurable this way or just one interface for all audio decoders?

Reply:
Won't libsnd give us what we want essentially for free - i.e. open not only ".wav", but also other supported formats?

Reply:
Agree. I changed this a little bit - named this libsnd adapter as `GenericDecoder` and it will serve as a fallback to all the formats.